clang/lib/Sema/SemaDecl.cpp

```diff
@@ -7416,10 +7416,11 @@ NamedDecl *Sema::ActOnVariableDeclarator(
     tryToFixVariablyModifiedVarType(TInfo, R, D.getIdentifierLoc(),
                                     /*DiagID=*/0);
 
-  if (const AutoType *AutoT = R->getAs<AutoType>())
-    CheckConstrainedAuto(
-        AutoT,
-        TInfo->getTypeLoc().getContainedAutoTypeLoc().getConceptNameLoc());
+  if (const AutoType *AutoT = R->getAs<AutoType>()) {
+    AutoTypeLoc Loc = TInfo->getTypeLoc().getContainedAutoTypeLoc();
```

Comment:
These two lines are analyzing the type in very different ways, so there seems to be further bugs here.

`Type::getAs<AutoType>` will simply desugar a type until it finds a node with type 'AutoType'.

`getContainedAutoTypeLoc()` will dig through a certain list of type nodes until it finds an 'AutoType'.
This list includes non-sugar nodes as well.

Let's assume for now we really want `R->getAs<AutoType>()` in the if condition here, which would not find the AutoType in `auto *`. This seems wrong, as these can be constrained as well. But we can handle that as a separate issue.

So to hit your bug, you are finding an AutoType through getAs, but not finding it through getContainedAutoTypeLoc().

My first suspicion would be that the type contains a sugar node not handled by `GetContainedAutoTypeLocVisitor`, in `AST/TypeLoc.cpp`.

Can you check that?

Reply:
Oh, looking at the reproducer, I see the problem:

Our variable is getting the type of an undeduced AutoType through that decltype calling a deleted function with auto type.

This is confusing our semantic analysis for the variable, as now we think the variable was declared as auto, as getAs will look through all sugar nodes, 'decltype' included.

Since `getContainedAutoTypeLoc()` doesn't look through decltype, you get a crash.

But we shouldn't be getting an 'auto' here in the first place.

The correct fix would be to make that decltype have some error type, for error recovery purposes.

Since we don't implement a generic unknown error type, 'int' would still be better.

Reply:
Thank you for the analysis!

> The correct fix would be to make that decltype have some error type, for error recovery purposes.

Can you point me out to the place in our source where this change should be made?

Reply:
Sure, but let me get back first to the other problem in this area:

I think that assumption I pointed out is wrong, we should not be using `getAs<AutoType>` here, as that will not dig through ReferenceType and PointerType, and the following example will not be diagnosed: https://godbolt.org/z/jzEaYPq4r

The `getContainedAutoTypeLoc` is what is a good fit for that, and what we should be using throughout. It will dig through the things needed for finding the AutoType, like pointers, references, and some of the necessary type sugar which can appear in source code when writing a deduced type. It does not dig through the things it doesn't need to, like `decltype`, because you can't write a deduced type with that.

So something like:
```suggestion
  if (AutoTypeLoc TL = TInfo->getTypeLoc().getContainedAutoTypeLoc()) {
    const AutoType *AT = TL.getTypePtr();
```

Should work, and it will not get confused by the error recovery issue, which can be addressed separately.

Nit: We conventionally use `TL` abbreviation for TypeLoc, and `Loc` is conventional for SourceLocation.

Reply:
Regarding the error recovery issue, in `SemaOverload.cpp`:

In `FixOverloadedFunctionReference`, for both `UnresolvedLookupExpr` and `UnresolvedMemberExpr`, we must avoid building a `DeclRefExpr` which has a function type returning undeduced auto.

Presently this should only happen for calls to deleted functions returning a deduced type, so should only happen when it's called from `FinishOverloadedCallExpr`, in the `OR_Deleted` case.

You can replace just the `auto` in the return type with `int`, keeping pointers and references in case of `auto *` or `auto &`. That can be done using `Sema::SubstAutoType`.

Eventually when we implement some kind of `ErrorTy`, for improving error recovery, we can change all such uses of 'int', using that instead.

Reply:
I applied your suggestion, leaving error recovery issue for later, and also added tests for pointer and reference types. Thank you again for your excellent analysis.

If that's how it should be fixed today, I'm fine proceeding this way. That said, I find the approach somewhat backwards: morally I'm interested whether there's a constrained auto that might be using a concept, but we implement it asking for _location_ of constrained auto, and get the AST node for constrained auto from it.

Reply:
But here we are not just asking for the location of constrained auto.

A TypeLoc is not just a SourceLocation, it's a QualType *with* source locations.

The reason you are starting with a TypeLoc is that you are also interested in the source location.

This is a bit confusing because in a lot of places we pass both a QualType and a TypeLoc, which is mostly redundant. That might induce folks into thinking that a TypeLoc is just source locations on the side, which is not true.

But presently, they are not always redundant though. There are some cases we store both separately, and apply some transformations, like type deduction, only to the QualType, and leave the TypeLoc alone.
I think this is not ideal though, and seems to be mostly due to tech debt.

As an aside, if all you need to do is mark the concepts as used, you might as well do that earlier when the AutoType is formed, for example around the calls to `ASTContext::getAutoType` in `SemaType.cpp` / `ConvertDeclSpecToType`.

Reply:
This makes more sense now.

> A TypeLoc is not just a SourceLocation, it's a QualType with source locations.

I though that `QualType` with source locations is `TypeSourceInfo`. If I correctly understand description of `TypeLoc` correctly, `TypeLoc` is trailing data of `TypeSourceInfo`. Is that correct?

> As an aside, if all you need to do is mark the concepts as used, you might as well do that earlier when the AutoType is formed, for example around the calls to ASTContext::getAutoType in SemaType.cpp / ConvertDeclSpecToType.

Calling `DiagnoseUseOfDecl` there seems too early. It triggers on _declaration_ of deprecated concept.

Reply:
Yeah, we don't need to have both TypeSourceInfo and TypeLoc in clang. I am not sure the history of how the situation arose, but we can pretty much consolidate on TypeLoc at some point.